### ISSUE TITLE
Sorting and identifing pages of students

### ISSUE THREAD

**Issue by Miguel-Hermanns** [sdaps/sdaps#203]
Hi,

we have a large number of students to grade (around 700) and we are using a mix of OMR sheets and normal paper sheets for it. We use the OMR sheets to get preliminary grades in a short period of time, and the students then decide if they want to review their full written exams (normal paper sheets) to correct/fine-tune the preliminary grade. 

For the time being, I sort manually the normal paper sheets alphabetically and then search for the exams of those students that want to review. Afterwards, I scan everything and with some speadsheet magic and shell scripts I cut the scanned pdf, put all parts of each student together, and send the output to them via email. As you can imagine, I'm interested in doing all this sorting and searching in an automatic way, and wonder if sdaps could help me in that.

What I need is to identify which pages of the scanned pdf belong to which student, and get as output a list of that matching. I envision the students filling in their names in each sheet using optical marks and then with sdaps I could know which student has filled in which page. Would that work? 

I have also seen the interesting feature of the barcodes and the questionaire-IDs, but I'm still unsure how to exploit that in an efficient, and tamper proof, way.

Thanks in advance for your help and best regards,

Miguel

**Comment by benzea** [comment]
> I envision the students filling in their names in each sheet using optical marks and then with sdaps I could know which student has filled in which page. Would that work?

I would suggest not doing that really. Filling in names is clumsy, cannot be processed automatically and is quite error prone in general. Even if you use e.g. bubbles for a student ID, you will still get recognition errors.

Yeah, I think using unique IDs is the solution. Either pre-assigning the IDs or assigning them on the fly during the exam. Differen strategies could be:

 * Use the questionnaire_id directly with a student ID (or a random ID that maps to the student). Also print e.g. the student name/id onto the sheet.
 * Use no IDs or random IDs (for multi-page safety); leave areas for a sticker and print a student-specific sticker. This scales even if students are seated randomly. With the "codebox" you can handle it inside SDAPS, or just do it outside using zbarimg directly.

**Comment by Miguel-Hermanns** [comment]
Why do you say that using bubbles for entering a student ID will still get recognition errors? Wouldn't it be exactly the same as recognizing the answers to a multiple choice question?

The use of unique IDs has its drawbacks (I think):

1) I cannot give to each student their unique generated set of normal paper sheets with their assigned questionaire_id, because the students enter the exam room in an unspecified order. One possibility could be to email them the normal paper sheet with their unique questionaire_id and ask them to print out as many copies as they assume they will need. But I'm 100% sure that many of them will come without the printed sheets, others will have not printed enough, etc. And then, to solve those issues in a reasonable way will be impossible.

2) I could handle to each student at the beginning of the exam a pile of normal paper sheets with a random ID. In that case, as long as sufficient sheets are given to them, the two issues pointed out above would no longer exist. But then I need to match that random ID with the corresponding student, which requires the student to fill out some kind of form that I can process automatically with sdaps. So we are again at the starting point, which is to identify which page (random ID) belongs to which student.

3) The option with the stickers is reasonable, but we don't have the infrastructure to generate and distribute the stickers to the students, neither before the exam nor during the exam. If the use of stickers would be adopted by all courses it would make sense. But only for our subject ...

**Comment by benzea** [comment]
> Why do you say that using bubbles for entering a student ID will still get recognition errors?

Well, I just assume that people do a bad job at actually filling it in. Though, even if a few students get it wrong, you can easily fix up those few cases manually (ask for student-id as bubble form + also let them enter name/id as a freeform field).

Printing stickers is pretty straight forward in principle. But yeah, it does require being able to tell the people who are in the room to hand them out properly … but it is the fool-proof way of making the students fill out some bubble form.

**Comment by Miguel-Hermanns** [comment]
In this case the students are used to fill out bubbles for student IDs, as many courses are using OMR sheets (but commercial ones). So I assume the rate of errors would be rather low. And yes, asking them to also put their names and student IDs allows a manual correction afterwards.

So, going back to the original question. Is it possible to use sdaps to identify which pages in a scanned pdf file correspond to which student? If yes, I see several challenges (or maybe not) that would need to be addressed:

1) Double-sided use of the sheets, while only one face of the sheet carries the marks.

2) The number of sheets of each student may vary.

3) Rarely get the sheets of different students mixed, but that would be no issue if a matching of each page with each student is possible.

4) When scanning the sheets, some may be wrongly oriented (flipped in any direction).

I don't know if sdaps is already able to handle all these challenges or if I would need to do something additional. Sorry, I have not been able to "play" with sdaps as I have still not managed to install it in Debian Buster :-(. I saw that there were issues with version 1.9.7, but 1.9.8 does not install either due to the same problem with python's version being 3.7.3.

**Comment by benzea** [comment]
Right, about those challenges:

1. Is not a problem in principle, SDAPS handles it
2. This is a bit of problem. You basically would need a "dummy" empty questionnaire with only two pages. SDAPS is stupid about this and just assume equal amount of pages each time.
3. Well, with 2. that would not matter …
4. This does not matter to SDAPS. However, I don't think we expose that information in a nice way.

Empty sheets are a bit of a problem really, because SDAPS would still want to have the corner marks for example. You don't need them, you would just need to convince SDAPS to export the information you need (and ignore all the warnings that happen). But at that point SDAPS is just a slightly smarter QR code reader …

**Comment by benzea** [comment]
Right, you want to put a bubble sheet for the student ID on each paper?

Well, in that case it is all self-contained. SDAPS would simply give you each digit of the student ID in the CSV file. As you get that for every double page, you can already pick out the correct images from that. Ideally you would also get the information about how the page is flipped/rotated.

**Comment by Miguel-Hermanns** [comment]
Yes, I would put a bubble sheet on each paper, otherwise it is impossible to do the matching, or?

Great to hear that everything seems to be possible. Now I only need to figure out how to install it on Debian Buster.

Thanks a lot for the help!

**Comment by Miguel-Hermanns** [comment]
Everything solved and running! I attach the generated questionnaire to give an idea of the result. I will try to run a small scale trial to also polish the workflow that comes after generating the csv, but I don't see any issues/difficulties/challenges.

[questionnaire.pdf](https://github.com/sdaps/sdaps/files/4117693/questionnaire.pdf)

**Comment by guillermoaguilar** [comment]
Hi Miguel, I'd be interested in seeing the source code for that file, it looks very nice and I'd like to use it in our next exam. Just the part you have there, the boxes for the names and the matrix for the student number. 
Thank you in advance!

**Comment by Miguel-Hermanns** [comment]
Hi Guillermo, find attached the LaTeX source file (I have changed the extension to TXT as otherwise Github does not allow me to upload it to this message):

[questionnaire.txt](https://github.com/sdaps/sdaps/files/4121089/questionnaire.txt)

I have given text and variable names to each field so that it is easier to identify the information in the csv file. I have also defined the boxes for the name and surname as variables to identify, although I think at the end I will eliminate them as I don't see a real use for it and I'm sure it will speed up the recognition process.

Once you have the digits of the student IDs imported into the spreadsheet you will need to combine them to get the real student IDs. I have not done that yet, so if you come up with clever ideas let me know :-). The problem I see is if the length of the student ID is less than 8 digits and the student starts to fill out the boxes from left to right, leaving the rightmost ones empty.

**Comment by benzea** [comment]
This is a bit obscure, but I believe you will get -1 or -2 as the result in the CSV. i.e. one of them meaning "nothing checked" and the other meaning "multiple items checked".

It should be pretty straight forward to put together the student ID with that information.

**Comment by benzea** [comment]
Hmm, just wondering. Did you intentionally use multiple `choicearray` environments to get the numbering in each row rather than only once?

**Comment by Miguel-Hermanns** [comment]
I have already seen the -1 output in my csv file. Good to know also about the -2. I agree, with these extra entries a postprocessing can be done to check for issues and construct the correct student ID.

I had three reasons for using multiple choicearrays instead of a single large one. The first one is precisely what you mention. To have numbers next to each column helps a lot to reduce errors while filling out the individual digits of the student ID. The second reason is that I was able to control much better the spacing between columns. By default the spacing is too large for what I wanted and I did not see an option to control that. And the third one is that I was able to force the numbers of the options to get squeezed against the bubbles, giving a much better (IMHO) output.

**Comment by benzea** [comment]
> The second reason is that I was able to control much better the spacing between columns.

Do you mean the spacing between number and box? Because there is are the `colsep=` and `rowsep=` optional arguments that should work nicely. Maybe that is what you were looking for?

**Comment by Miguel-Hermanns** [comment]
Did not see those arguments. I have just played with them and I get a strange behaviour. See for example the following code. By changing the colsep I get inconsistent results. Sometimes by putting 1mm everything works as expected, other times the numbers disappear behind the bubbles, other times the disappear completely, etc.:

`    \small
      \begin{choicearray}[vertical,singlechoice,rowsep=0mm,colsep=2mm]{}
        \choice[val=0]{0}
        \choice[val=1]{1}
        \choice[val=2]{2}
        \choice[val=3]{3}
        \choice[val=4]{4}
        \choice[val=5]{5}
        \choice[val=6]{6}
        \choice[val=7]{7}
        \choice[val=8]{8}
        \choice[val=9]{9}
        \question[text=ID1,var=ID1]{A}
        \question[text=ID2,var=ID2]{B}
      \end{choicearray}
`
Indeed, with these arguments it would be possible to control the spacing between rows and columns. It would also allow the squeezing of the numbers against the first column of bubbles.

**Comment by benzea** [comment]
Did you recompile multiple times? The array environment there requires two passes to stabilize the layout, so odd things are likely to happen if you just compile once.

**Comment by Miguel-Hermanns** [comment]
You are right. That was the issue. When I compile twice everything behaves as expected. Many thanks.

**Comment by Miguel-Hermanns** [comment]
Just wanted to update information on this use case. It has gone absolutely flawless! I have been able to send by email the written exams to over 500 students without any problems. We are talking about a total of over 6000 paper sheets (questionnaires) that have been processed, identified, chopped, put together, and sent by email.

In this task, the combination of an excel sheet checking everything and warning about the odd cases (double marks, nonexistent students, unknown pages, unfilled questionnaires, etc) combined with sdaps’s GUI to correct these odd cases (mostly wrongly marked ID numbers) has been extremely useful and effective.

So, thanks a lot to all of you developing sdaps!
